_overviews/hello-scala/for-expressions.md

```diff
@@ -0,0 +1,144 @@
+
+For example, given this list of integers:
+
+```scala
+val nums = Seq(1,2,3)
+```
+
+You can create a new list of integers where all of the values are doubled, like this:
+
+```scala
+val doubledNums = for (n <- nums) yield n * 2
+```
+
+That expression can be read as, “For every number `n` in the list of numbers `nums`, double each value, and then assign all of the new values to the variable `doubledNums`.” This is what it looks like in the Scala REPL:
+
+```scala
+scala> val doubledNums = for (n <- nums) yield n * 2
+doubledNums: Seq[Int] = List(2, 4, 6)
+```
+
+As the REPL output shows, the new list `doubledNums` contains these values:
+
+```scala
+List(2,4,6)
+```
+
+In summary, the result of the for-expression is that it creates a new variable named `doubledNums` whose values were created by doubling each value in the original list, `nums`.
+
+
+
+## Capitalizing a list of strings
+
+You can use the same approach with a list of strings. For example, given this list of lowercase strings:
+
+```scala
+val names = List("adam", "david", "frank")
+```
+
+You can create a list of capitalized strings with this for-expression:
+
+```scala
+val ucNames = for (name <- names) yield name.capitalize
+```
+
+The REPL shows how this works:
+
+```scala
+scala> val ucNames = for (name <- names) yield name.capitalize
+ucNames: List[String] = List(Adam, David, Frank)
+```
+
+Success! Each name in the new variable `ucNames` is capitalized.
+
+
+
+## The `yield` keyword
+
+Notice that both of those for-expressions use the `yield` keyword:
+
+```scala
+val doubledNums = for (n <- nums) yield n * 2
+                                  -----
+
+val ucNames = for (name <- names) yield name.capitalize
+                                  -----
+```
+
+Using `yield` after `for` is the “secret sauce” that says, “I want to yield a new collection from the existing collection that I’m iterating over in the for-expression, using the algorithm shown.”
+
+
+
+## Using a block of code after `yield`
+
+The code after the `yield` expression can be as long as necessary to solve the current problem. For example, given a list of strings like this:
+
+```scala
+val names = List("_adam", "_david", "_frank")
+```
+
+Imagine that you want to create a new list that has the capitalized names of each person. To do that, you first need to remove the underscore character at the beginning of each name, and then capitalize each name. To remove the underscore from each name, you call the `tail` method on each `String`. After you do that, you call the `capitalize` method on each string. Here’s how you can use a for-expression to solve this problem:
+
+```scala
+val capNames = for (name <- names) yield {
+    val nameWithoutUnderscore = name.tail
```

Comment:
Do you prefer `drop(1)` because it's more easy to understand than `tail` for beginners?

Reply:
No, only because I think it’s safer to first introduce total operations. Every line of code that we put here will be kind of promoted to a “standard way of doing things”. I believe that using `tail` is not trivial because it is partially defined (and there are no checked exceptions in Scala) and this method should be put into the hands of experts rather than beginners.